src/google/protobuf/compiler/ruby/ruby_generator.cc

```diff
@@ -451,8 +466,18 @@ bool GenerateFile(const FileDescriptor* file, io::Printer* printer,
     }
   }
 
+  printer->Print("Google::Protobuf::DescriptorPool.generated_pool.build(\n");
+  printer->Indent();
+  printer->Print("Google::Protobuf::FileDescriptor.new.tap do |f|\n");
```

Comment:
Maybe the syntax could be a parameter to the constructor? The syntax is a fundamental property of a file, and in an actual .proto file it has to be the first statement.

Making the syntax a constructor parameter and then defining all messages/enums inside the block makes sense to me.

If we are planning to go farther down the DSL route, we'll be wanting to add the ability to specify options of files, messages, enums, etc. To do that we'll need to mirror the structure of actual proto descriptors as closely as possible. Files contain messages, and every message belongs to a file.

Reply:
Seems reasonable, but could you clarify what you mean by "constructor parameter", so something like (super rough):

```ruby
require 'other_pb.rb"

Google::Protobuf::FileDescriptor.start do
  syntax :proto2
  option :java_package,  "com.example.foo"
 
  pool.build do
    add_message ...
  end
end
```

Reply:
I was thinking more along the lines of:

```ruby
Google::Protobuf::DescriptorPool.generated_pool.build do
  add_file(syntax => :proto2) do
    add_message ...
  end
end
```

I am not exactly sure yet how we should specify options. The problem is that options can get *really* tricky. You can define custom options that use your own types. But even weirder, you can *use* the option on the same type that is *defining* the option:

```proto
syntax = "proto2";

import "google/protobuf/descriptor.proto";

package pkg;

message M {
  // This will become an option on message M, but its value is also
  // an instance of the M message.  This is really tricky because it
  // implies that you need to be able to create an M message while
  // you are defining it.
  //
  // The protobuf library generally deals with this by encoding options
  // as serialized protos, and only parsing them once the message
  // types are defined.
  option (pkg.m) = {val: 999};
  optional int32 val = 1;
}

extend google.protobuf.MessageOptions {
  optional M m = 1234567;
}
```

Whatever path we go down with options, I want to make sure it can accommodate the trickiest cases, like the above. This is one of the reasons the DSL worries me: it's easy to design yourself into a corner if you're not careful.

Reply:
Gotcha, that is fun to deal with. Do you have an example somewhere of what it looks like in another language?

Reply:
You can try compiling my snippet above into other languages to see how it is handled. Python might be a good model since it's similar to Ruby. Python defines the types, then registers extensions (including custom options), then sets options again. The options are specified as a serialized binary string, which will find any extensions that are registered:

```python
_M = _descriptor.Descriptor(
  name='M',
  full_name='pkg.M',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='val', full_name='pkg.M.val', index=0,
      number=1, type=5, cpp_type=1, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      options=None),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  options=_descriptor._ParseOptions(descriptor_pb2.MessageOptions(), _b('\272\350\332\004\003\010\347\007')),
  is_extendable=False,
  syntax='proto2',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=53,
  serialized_end=79,
)

DESCRIPTOR.message_types_by_name['M'] = _M
DESCRIPTOR.extensions_by_name['m'] = m

M = _reflection.GeneratedProtocolMessageType('M', (_message.Message,), dict(
  DESCRIPTOR = _M,
  __module__ = 'test_pb2'
  # @@protoc_insertion_point(class_scope:pkg.M)
  ))
_sym_db.RegisterMessage(M)

m.message_type = _M
google_dot_protobuf_dot_descriptor__pb2.MessageOptions.RegisterExtension(m)

_M.has_options = True
_M._options = _descriptor._ParseOptions(descriptor_pb2.MessageOptions(), _b('\272\350\332\004\003\010\347\007'))
```

This approach could work for Ruby. We should freeze the descriptor after we set options, so users don't try to add or mutate options on the descriptor after it has been built.